lib/inets/test/httpd_serve_SUITE.erl

```diff
@@ -0,0 +1,241 @@
+    HttpcOpts = [{socket_opts, [{ipfamily, inet6fb4}]}],
+    Request = {Url, []},
+    Response = httpc:request(get, Request, [{autoredirect, false}], HttpcOpts),
+    case Response of
+        {ok, {{_Version, 200, _}, _Headers, _Body}} ->
+            Response;
+        {ok, {{_Version, 301, _}, Headers, _Body}} ->
+            % To be resilient against the case where the server is not
+            % reachable under its `server_name`, for instance, in test
+            % containers or other hosts with unexpected networking setups,
+            % replace the suggested hostname with the hostname we came from.
+            {_, SuggestedTarget} = proplists:lookup("location", Headers),
+            #{path := SuggestedPath} = uri_string:parse(SuggestedTarget),
+            OurUri = uri_string:parse(Url),
+            DecomposedTarget = maps:put(path, SuggestedPath, OurUri),
+            RedirectTarget = uri_string:recompose(DecomposedTarget),
+            RedirectedRequest = {RedirectTarget, []},
+            RedirectedResponse = httpc:request(get, RedirectedRequest, [], HttpcOpts),
+            ct:log("Following redirect (rewritten from ~s to ~s)", [SuggestedTarget, RedirectTarget]),
+            {ok, {{_, 200, _}, _, _}} = RedirectedResponse;
+        {error, {failed_connect, [{to_address, {"::", _Port}},
+                                  {inet6, [inet6], eaddrnotavail},
+                                  {inet, [inet], nxdomain}]}} ->
+            % In this case we could bind on all IPv6 interfaces (::) just
+            % fine, but could not issue requests to it due to an OS error.
+            % Write it off as a networking misconfiguration and skip.
+            exit({skip, "Could not reach host on IPv6 address ::"})
+    end.
+
+verify_assertions([], _ServerInfo) ->
+    ok;
+
+verify_assertions([directory_index | Assertions], #{url := Url} = ServerInfo) ->
+    ct:log("Validating directory index at ~s", [Url]),
+    verify_200_at(Url),
+    ct:log("Directory index received with a 200"),
+    verify_assertions(Assertions, ServerInfo);
+
+verify_assertions([random_file | Assertions],  #{url := Url, path := Path} = ServerInfo) ->
+    Files = filelib:wildcard(Path ++ "/*"),
+    File = lists:nth(rand:uniform(length(Files)), Files),
+    Basename = filename:basename(File),
+    FileUrl = Url ++ Basename,
+    ct:log("Validating random file at ~s", [FileUrl]),
+    verify_200_at(FileUrl),
+    ct:log("File received with a 200"),
+    verify_assertions(Assertions, ServerInfo).
+
+%%
+%% Helper functions
+%%
+
+suite_data(Config) ->
+    proplists:get_value(data_dir, Config).
+
+verify_server(FunOrArgs) ->
+    TestFun = fun run_server_assertions/1,
+    with_server(FunOrArgs, TestFun).
+
+with_server(Args, TestFun) when is_list(Args) ->
+    ServerFun = fun () -> httpd:serve(Args) end,
+    run_with_server(TestFun, ServerFun);
+
+with_server(ServerFun, TestFun) when is_function(ServerFun) ->
+    run_with_server(TestFun, ServerFun).
```

Comment:
maybe I'm missing some convention, but I don't see point in change arguments order between `with_server` and `run_with_server`. wouldn't it be simpler is both functions take ServerFun as 1st and TestFun as 2nd argument?